core/src/main/java/org/apache/carbondata/core/scan/complextypes/ComplexQueryType.java

```diff
@@ -67,4 +67,18 @@ private DimensionColumnPage getDecodedDimensionPage(DimensionColumnPage[][] dime
     }
     return dimensionColumnPages[columnIndex][pageNumber];
   }
+
+  /**
+   * Method will copy the block chunk holder data and return the cloned value.
+   * This method is also used by child.
+   */
+  protected byte[] copyBlockDataChunkWithoutClone(DimensionRawColumnChunk[] rawColumnChunks,
+      DimensionColumnPage[][] dimensionColumnPages, int rowNumber, int pageNumber) {
+    byte[] data =
+        getDecodedDimensionPage(dimensionColumnPages, rawColumnChunks[columnIndex], pageNumber)
```

Comment:
how about to cache the page, it will not require to decode for each row again.

Reply:
I have debugged, cache is already there. the argument of this method, `DimensionColumnPage[][] dimensionColumnPages` itself is a cache based on column index.

go inside `ComplexQueryType#getDecodedDimensionPage` to see it.

Also observed that only once decodeColumnPage called for that page, reset it is using from cache only.

Reply:
In BlockletScannedResult, dimensionColumnPages[][]